pkg/querier/tenantfederation/merge_queryable.go

```diff
@@ -0,0 +1,309 @@
+
+}
+
+type stringSliceFunc func(storage.Querier) ([]string, storage.Warnings, error)
+
+// mergeDistinctStringSlice is aggregating results from stringSliceFunc calls
+// on a querier. It removes duplicates and sorts the result.
+//
+// TODO: Consider running stringSliceFunc calls concurrently
+func (m *mergeQuerier) mergeDistinctStringSlice(f stringSliceFunc) ([]string, storage.Warnings, error) {
+	var warnings storage.Warnings
+	resultMap := make(map[string]struct{})
+	for pos, tenantID := range m.tenantIDs {
+		result, resultWarnings, err := f(m.queriers[pos])
+		if err != nil {
+			return nil, nil, err
+		}
+		for _, e := range result {
+			resultMap[e] = struct{}{}
+		}
+		for _, w := range resultWarnings {
+			warnings = append(warnings, fmt.Errorf("error querying tenant id %s: %w", tenantID, w))
+		}
+	}
+
+	var result []string
+	for e := range resultMap {
+		result = append(result, e)
+	}
+	sort.Strings(result)
+	return result, warnings, nil
+}
+
+// Close releases the resources of the Querier.
+func (m *mergeQuerier) Close() error {
+	var errs prometheus.MultiError
+	for pos, tenantID := range m.tenantIDs {
+		if err := m.queriers[pos].Close(); err != nil {
+			errs = append(errs, fmt.Errorf("failed to close querier for tenant id %s: %w", tenantID, err))
+		}
+	}
+	return errs.MaybeUnwrap()
+}
+
+// Select returns a set of series that matches the given label matchers. If the
+// tenantLabelName is matched on it only considers those queriers matching. The
+// forwaded labelSelector is not containing those that operate on tenantLabelName.
+func (m *mergeQuerier) Select(sortSeries bool, hints *storage.SelectHints, matchers ...*labels.Matcher) storage.SeriesSet {
+	tenantIDsPos, filteredMatchers := filterValuesByMatchers(string(m.tenantLabelName), m.tenantIDs, matchers...)
+	var seriesSets = make([]storage.SeriesSet, len(tenantIDsPos))
+	for pos, posTenant := range tenantIDsPos {
+		tenantID := m.tenantIDs[posTenant]
+		seriesSets[pos] = &addLabelsSeriesSet{
+			// TODO: Consider running Select calls concurrently
+			upstream: m.queriers[posTenant].Select(sortSeries, hints, filteredMatchers...),
+			labels: labels.Labels{
+				{
+					Name:  string(m.tenantLabelName),
+					Value: tenantID,
+				},
+			},
+		}
+	}
+	return storage.NewMergeSeriesSet(seriesSets, storage.ChainedSeriesMerge)
+}
+
+// filterValuesByMatchers applies matchers to inputed labelName and labelValue.
+// The index of matched values is returned and also all label matcher not
+// matching the labelName.
+// In case a label matcher is set on a label confliciting with tenantLabelName,
+// we need to rename this labelMatcher to it's original name.
+// This is used to as part of Select in the mergeQueryable, to ensure only
+// relevant queries are considered and the forwarded matchers do not contain
+// matchers on the tenantLabelName.
+func filterValuesByMatchers(labelName string, labelValues []string, matchers ...*labels.Matcher) ([]int, []*labels.Matcher) {
```

Comment:
I would name returned parameters in this case, just for clarity:

```suggestion
func filterValuesByMatchers(labelName string, labelValues []string, matchers ...*labels.Matcher) (matchedValuesPos []int, unrelatedMatchers []*labels.Matcher) {
```

Reply:
Good point done in 4e1c8900bf9bfc7fb33a7f0cfbbb32e8ad567662